### PR TITLE
RavenDB-18561 Add id to failed document replication error

### PR DESCRIPTION
**Pull request by haludi**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-18561

### Additional description
Log the document id in case of replication failur

### Type of change
- Additional error information

### How risky is the change?
- Low 

### Backward compatibility
- Not relevant

### Is it platform specific issue?
- No

### Documentation update
- No documentation update is needed 

### Testing 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?
- No

### UI work
- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 1]
```diff
@@ -1240,146 +1240,153 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
...
-                                        throw new ArgumentOutOfRangeException(nameof(conflictStatus),
-                                            "Invalid ConflictStatus: " + conflictStatus);
+                                        default:
+                                            throw new ArgumentOutOfRangeException(nameof(conflictStatus),
+                                                "Invalid ConflictStatus: " + conflictStatus);
+                                    }
+                                }
+                                catch (Exception e)
+                                {
+                                    throw new InvalidOperationException($"Doc id:\"{doc.Id}\"", e);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]:
```suggestion
                                    throw new InvalidOperationException($"Failed to replicate document: {doc.Id}", e);
```

Also, what about failures for attachments, counters, time series, tombstones, etc?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 2]
```diff
@@ -1108,284 +1108,292 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
...
-                            default:
-                                throw new ArgumentOutOfRangeException(item.GetType().ToString());
+                                default:
+                                    throw new ArgumentOutOfRangeException(item.GetType().ToString());
+                            }
+                        }
+                        catch (Exception e)
+                        {
+                            var errorMessage = BuildErrorMessage(item);
+                            throw new InvalidOperationException(errorMessage, e);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: I don't like that we change the type of the exception here... maybe add this to the `Data`?

Comment (haludi) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: But then we will not see it in the logs.
I can just log it but then we will see this error twice in the logs (not so bad).
Or I can implement the logs to also write the data.

Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: 1. We need to see that in the logs
2. We need to see that in the notification center as well

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 3]
```diff
@@ -1422,6 +1448,23 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
...
+                {
+                    AttachmentReplicationItem item => item.Key.ToString(),
+                    AttachmentTombstoneReplicationItem item => item.Key.ToString(),
+                    CounterReplicationItem item => item.Id,
+                    DocumentReplicationItem item => item.Id,
+                    RevisionTombstoneReplicationItem item => item.Id,
+                    TimeSeriesDeletedRangeItem item => item.Key.ToString(),
+                    TimeSeriesReplicationItem item => item.Key.ToString(),
+                    _ => throw new InvalidOperationException($"Type {baseItem.Type} is unknown")
+                };
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1463]: please use `DocumentInfoHelper.GetItemInformation`

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 4]
```diff
@@ -1108,284 +1112,306 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
...
+                            _replicationInfo.NotificationCenter.Add(
+                                AlertRaised.Create(
+                                    database.Name,
+                                    "Replication Failure",
+                                    errorMessage,
+                                    AlertType.Replication,
+                                    NotificationSeverity.Error, 
+                                    key:$"{_replicationInfo.SourceDatabaseId}:{_mode}",
+                                    details: new ExceptionDetails(e)
+                                ));
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1413]: Upon successful replication we need to dissmis this notification.

Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1413]: Upon successful replication we need to dissmis this notification.

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 5]
```diff
@@ -1108,284 +1112,306 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
...
+                            }
+                        }
+                        catch (Exception e)
+                        {
+                            var errorMessage = BuildErrorMessage(item);
+                            
+                            if (_replicationInfo.Logger.IsOperationsEnabled)
+                                _replicationInfo.Logger.Operations(errorMessage, e);
+                            
+                            _replicationInfo.NotificationCenter.Add(
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1404]: use `database.NotificationCenter` ?

Comment (karmeli87) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:1404]: use `database.NotificationCenter` ?

**Comment by ppekrol** [comment]
Please reopen when work will be done.

**Pull request closed without merging**